src/ucp/tag/tag_recv.c

```diff
@@ -101,7 +101,12 @@ ucp_tag_recv_common(ucp_worker_h worker, void *buffer, size_t count,
     req->recv.tag.tag_mask  = tag_mask;
     if (param->op_attr_mask & UCP_OP_ATTR_FIELD_CALLBACK) {
         req->recv.tag.cb    = param->cb.recv;
-        req->user_data      = param->user_data;
+
+        if (param->op_attr_mask & UCP_OP_ATTR_FIELD_USER_DATA) {
```

Comment:
maybe move this assignment outside of `if (param->op_attr_mask & UCP_OP_ATTR_FIELD_CALLBACK)` statement?

Reply:
But user_data is needed / used only if callback is set.

Reply:
ah, I see